### DISCUSSION TITLE
Should rollback the base64 encoding changes for HTTP Basic Auth?

### DISCUSSION THREAD

**Discussion by lexcao** [keycloak/keycloak#15427] [Ideas]
### Background

This PR https://github.com/keycloak/keycloak/pull/14179 introduced changes on base64 encoding for HTTP basic header, 
which made the default implementation to compliant with [RFC 2617](https://datatracker.ietf.org/doc/html/rfc2617).

### Changes 
```diff
- return "Basic " + Base64.encodeBytes((username + ':' + password).getBytes(StandardCharsets.UTF_8));
+ return "Basic " + Base64.encodeBytes((username + ':' + password).getBytes(StandardCharsets.UTF_8), Base64.DO_BREAK_LINES);
```

### Tracked issues
I have tracked some issues that maybe be caused by these changes.

* https://github.com/keycloak/keycloak/issues/14441
* https://github.com/keycloak/keycloak/discussions/15426
* https://github.com/keycloak/keycloak/issues/14530
* https://github.com/keycloak/keycloak/discussions/14500
* https://github.com/keycloak/keycloak/issues/14529
* https://github.com/keycloak/keycloak/discussions/15426

---
Need the Keycloak team to check.
WDYT, @stianst @mposolda

**Comment by sschu** [comment]
@lexcao We also have problems due to this change - but it is still IMHO correct to be compliant to the spec, don't you think? Or why would you like to roll back the change?

**Reply by lexcao** [reply]
Yes, it is compliant to the spec. I agree with you.

I am thinking if it needs to roll back because there are some issues reported.

TBH, I don't know how to deal with this case.
From the reaction (no response) from the Keycloak team, I think it may keep this change?

**Reply by sschu** [reply]
I'll try to get some feedback. Will keep you posted.

**Reply by jbman** [reply]
A test case which fails with new base64 decoding is a Client ID "Aladdin??" and any secret.
For example encoding "Aladdin??:open sesame" encoded with
* base64 results in QWxhZGRpbj8/Om9wZW4gc2VzYW1l
* base64Url results in QWxhZGRpbj8_Om9wZW4gc2VzYW1l

When decoding the header the kind of encoding could be detected based on  '-' and '_' or being present. If this is the case base64Url could be used. If the characters are not present base64.

**Reply by mposolda** [reply]
@lexcao Thanks for tracking the related issues.

@lexcao @sschu @jbman +1 that we should be compliant with the specs. So if current behaviour is compliant with the specs, we should not change it.

However what we possibly missed is to document this change in the "Upgrading guide", which we we did not and it is probably something, which we should have done due this is kind of change, which breaks the existing behaviour. Even if existing behaviour (before applying this change) relied on the non-specs compliant behaviour. Anyone is welcome to send PR to the "Upgrading guide" for add something about this.

**Reply by sschu** [reply]
@mposolda We just discussed with @stianst  that we will provide a PR to implement the compatible behaviour @jbman suggested. This should even be included in 20.0.2. Since one of our customers stumbled across this issue and he is using Spring, apparently even the Spring client libraries use the wrong encoding.

**Reply by sschu** [reply]
After trying to implement this I realised the automatic detection might be a bit more complicated since you cannot really decide if the old or new encoding is used. For encoded values that neither contain "/","+" (Base64) or "-","_" (Base64URL), both the old and the new way could have been used. 

In the old way, this would just be client_id and secret and in the new way, both would have to be URLdecoded first. This is not a problem for the secret since Keycloak generates secrets to not contain URL-unsafe characters - so encoded and unencoded value is the same. For the client id, that is a problem since client ids can contain arbitrary characters.

If the value received is not URL-safe, then it was not encoded so it can directly be used as client id.
If the value received is URL-safe and does not contain any % characters, it does not matter whether it was encoded or not as both values are the same.
If the received value is URL-safe and contains % characters, there is no safe way to say whether is was URL-encoded e.g. "%2F" could be part of the client id but it could also have been "/".

The only safe way I see for that case is to actually look up whether the client exists in Keycloak.  WDYT @mposolda ?

**Reply by jbman** [reply]
I missed the problem that the client ID and secret each of them are expected to be URL encoded by the client. This makes detection not work.

Another option is complete new `AbstractClientAuthenticator`. Maybe this is the cleanest option, as the new implementation can be used for new clients, while others have the existing one configured.
In Admin Console the existing client authenticator has name "Client Id and Secret". This could be reverted as it was before. A new one would be named "Client ID and Secret (OAuth 2.0 compliant)" and contains the implementation as it is now.

**Reply by jbman** [reply]
This could even be used to introduce a Basic-Auth-only-option for clients, as RFC6749 says Basic Auth should be preferred over POST.

Would need two new client authenticators (based on same class):

- Client ID and Secret via Basic Auth
- Client ID and Sceret via Basic Auth or POST (new default)

**Comment by u3r** [comment]
A colleague of mine did some digging and we discussed this:

**Resources:**
Obsoleted Basic Auth RFC referenced by OAuth2 RFC: https://www.rfc-editor.org/rfc/rfc2617

Current Basic Auth RFC: https://www.rfc-editor.org/rfc/rfc7617

OAuth2 RFC section about Basic Auth for client authentication using password: https://datatracker.ietf.org/doc/html/rfc6749#section-2.3.1

**There are to Errata referencing this:**
https://www.rfc-editor.org/errata/eid6017 proposing to drop the section about applying application/x-www-form-urlencoded which apparently is interpreted as applying urlencoding to the parts of the basic auth's username and password. _"RFC 2617 has been superseded by RFC7617 which clearly defines in section 2.1 how a charset can be provided to solve the usecase described with encoding. The original text of this RFC violates the approach described for Basic authentication."_

https://www.rfc-editor.org/errata/eid4749 by Phil Hunt (co-author of the original RFC) proposes to make the intention of 2-step / double encoding (urlencode + base64) explicit to resolve the conflict with RFC2617 basic auth.

These errata are **reported** but neither **rejected** nor **verified**

**Opinion**
As to the RFC:
Personally we support the first errata entry (eid6017) which would result in a rollback of the mentioned change as we think
a) most clients currently do only standard basic-auth without double-encoding
b) basic auth is well established and thin obvious thing to do
c) double encoding seems not to provide measurable benefits in allowable characters for the username + password (only for very uncommon edge cases)

Keycloak code-wise it seems unlikely (if not impossible) to auto-detekt the method used by the client.
So alternatives are:
i) Keep the new implementation
ii) Roll back to previous state with no "extras"
iii) Make the behaviour configurable on Keycloak-installation level (with one or the other interpretation being the default)
iiii) Make the behaviour client selectable (with one or the other interpretation being the default)

Here ii or iii would be most attractive for us, as 
i causes more problems than it solves
and iiii introduces a custom extension of the standard - again splitting users and communities.

**Reply by sschu** [reply]
Thanks for your analysis @u3r. However, eid6017 would not result in a rollback of the changes because according to the errata the encoding should still be Base64 and not Base64URL.
Personally, my favourite solution would also be to just update the reference in RFC6749 to reference RFC7617 instead of the superseded RFC2617 and I am quite surprised this has not happened.

However, as it stands only the new implementation is in line with what the current spec (RFC6749) says, so I don't think removing it is an option. (Option ii)
I also don't like option iii) as it makes it very hard to migrate clients one after another.
Option iiii) would be valid, but I would still rather try to auto-detect this. I think this would be possible if we actually try to look up a client while decoding for the final decision.

@mposolda WDYT?

**Reply by u3r** [reply]
@sschu Ok, maybe not full rollback then.
So what should be the correct solution - or maybe rather the most compatible version (based on what current client implementations are doing)?
Only showing encoding (as far as I understood):
1) `base64(clientId + ":" + clientSecret)`  -> pure basic auth standard
2) `base64Url(clientId + ":" + clientSecret)` -> original KC implementation
3) `base64(urlEncode(clientId) + ":" + urlEncode(clientSecret))` -> the maximum complexity taken from Oauth spec

**Reply by mposolda** [reply]
@sschu @u3r I think option iiii is fine and we can have this configurable possibly? In the keycloak admin console, there is section `Open ID Connect Compatibility Modes` in the client settings (in the new admin console when you click to some client and then tab `Advanced`). We can possibly have the compatibility switch added here?

Regarding auto-detection, it can be nice but it depends if you are able to do it in a reliable way? Considering various corner cases (For example `client_id` containing only ASCII characters like `myclient`, but `client_secret` of this client containing some strange characters, etc).

**Reply by sschu** [reply]
@mposolda Since client secrets are generated by Keycloak, they only contain URL-safe characters. The only problem is client names. If they contain a string %XY, it is unclear whether it was encoded or whether this is really part of the client id (also see my full analysis above). The question is if it is fine to just look up the client id and check if the client exists (i.e. first check if client exists URL-decoded and if not check if it exists without decoding).

**Reply by mposolda** [reply]
@sschu Not sure. I wonder if this doesn't add possibility for some kind of attacks for example? Like someone registers the client with similar name like yours, but just with encoded characters. Then incorrect client might be looked-up during client authentication etc.

Also the client secrets are unfortunately not always generated - there is possibility to assign them (at least during realm import, but maybe also by some other means...).

**Reply by sschu** [reply]
@mposolda True, this might be possible. Then maybe the explicit way with compatibility settings is the better one. Would also make it clear for users that there is a compatibility switch which might be deprecated in the future.

**Reply by mposolda** [reply]
+1

**Reply by sschu** [reply]
@mposolda I just realised there still needs to be some auto-detection magic because you have to decode the client before you can retrieve its compatibility settings. ;) I will try to create a PR for this.

**Reply by sschu** [reply]
@mposolda There is one last ambiguity that cannot be resolved. Assume there is a client with client_id `/` and a client with client_id `%2F` that has compatibility mode enabled. The first will be encoded RFC-compliant as `JTJG`, the latter has the same encoded value using the compatible encoding. Keycloak would have to make a decision which one to choose.  So even with explicit settings per client, there is a possibility for a client not being able to authenticate if Keycloak selects the wrong one... I am not sure how relevant this problem is, I would at least assume that `%XY`is not commonly used.

**Reply by mposolda** [reply]
@sschu I wonder if it is possible to throw exception in case that such clients exists? As I don't see why someone want to use both cases in the valid deployment?

We can possibly also reject create/update client `%2F` in case that client `/` exists and vice-versa.
